### PR TITLE
Feature/async choices 2

### PR DESCRIPTION
**Pull request by davecoates**

### CODE REVIEW

File: .isort.cfg
Comment (levic) [inline .isort.cfg]: This doesn't seem to be referenced anywhere in the diff -- was it a bug fix or an accidental commit?

File: django-testbed/django-root/xenopus_frog/views.py
Comment (levic) [inline django-testbed/django-root/xenopus_frog/views.py]: For consistency that should be the first field listed

File: .isort.cfg
Comment (davecoates) [reply .isort.cfg]: was causing CI to fail - has since been fixed in master

File: js-packages/@prestojs/ui-antd/src/getWidgetForField.ts
Comment (levic) [inline js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:52]: This is `(field.choices || field.asyncChoices) ? ... : `? -- it might be worth putting brackets around that to make it a bit clearer what the order of precedence is (or maybe it's just me not seeing enough of these JS patterns)

File: js-packages/@prestojs/ui-antd/src/getWidgetForField.ts
Comment (levic) [inline js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:53]: This is `(field.choices || field.asyncChoices) ? ... : ... `? -- it might be worth putting brackets on line 50 to make it a bit clearer what the order of precedence is (or maybe it's just me not seeing enough of these JS patterns). Assuming prettification doesn't remove them.

File: js-packages/@prestojs/ui-antd/src/getWidgetForField.ts
Comment (levic) [inline js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:80]: As above re parentheses

File: js-packages/@prestojs/util/src/comparison.ts
Comment (levic) [inline js-packages/@prestojs/util/src/comparison.ts]: Grammar there is off

File: django-testbed/frontend/src/models/User.js
Comment (levic) [inline django-testbed/frontend/src/models/User.js]: For a normal app you'd presumably either know whether or not pagination was going to be present on the backend.

Is this something that should perhaps be rolled into presto core and printed as a console warning in dev?

File: django-testbed/frontend/src/models/User.js
Comment (levic) [inline django-testbed/frontend/src/models/User.js:35]: The `endpoints` definition above we'd put in the code generator?

File: js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx
Comment (levic) [inline js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx:18]: That link is broken.

File: js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx
Comment (levic) [inline js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx:256]: Comments re link & ant coupling as above

File: js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx
Comment (levic) [reply js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx:18]: How tightly coupled is this to ant?

More specifically, this is the antd widget implementation so I would expect it to be highly dependent on ant, but if I wanted to use another widget lib then could I potentially to see a quite different async choices interface?

File: django-testbed/frontend/src/models/User.js
Comment (levic) [inline django-testbed/frontend/src/models/User.js:106]: There seems to be a lot more boilerplate than I would have expected for this -- is this all required or is it testing every option?

Do you have a minimal example for an async choices that would be a lookup into another table? eg a list of Organisations, perhaps filtered by what the user is allowed to see.

File: js-packages/@prestojs/util/src/Labeled.ts
Comment (levic) [inline js-packages/@prestojs/util/src/Labeled.ts:3]: Am I able to read AsyncChoices documentation or only after it's merged?

File: js-packages/@prestojs/util/src/Labeled.ts
Comment (levic) [inline js-packages/@prestojs/util/src/Labeled.ts]: Is Plain vs Rich used elsewhere in react?

It makes sense but it's just terminology I haven't seen before. (TextLabelled vs NodeLabelled maybe)?

Beyond that, if a string and a Node are both renderable by React why do we need the distinction?

Comment (levic) [reply js-packages/@prestojs/util/src/Labeled.ts]: Edit: is this because some ant components insist on strings instead of renderable nodes?

File: js-packages/@prestojs/util/src/identifiable.ts
Comment (levic) [inline js-packages/@prestojs/util/src/identifiable.ts]: Why does this exist? If we already have `IdentifiablePk` then shouldn't that be the thing used to access the PK, and `id` is just the site-specific implementation of using `id` as the concrete PK?

If another backend was to use `Id` or `MyModel_Id` then wouldn't that require changes to presto?

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:61]: Can you give an example here of what a "dependency" is? How does it differ from the actual data itself?

eg if I have a list of organisations & their countries then is this supposed to query the country list or is that supposed to be a join that is going as part of the data returned by `list()` ?

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:63]: `params` is where an autocomplete keyword search string would fit in? How would that get injected given that `useListProps()`is passed straight in?

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:86]: Same comment as `useListProps` above

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:98]: Same comment as `list` above

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:119]: Is this intended to handle `null` values too or would the Widget handle that?

**Review by levic — CHANGES_REQUESTED** [review]

File: django-testbed/frontend/src/models/User.js
Comment (davecoates) [reply django-testbed/frontend/src/models/User.js]: This is obsoleted by #68 which removes this check (because it's dependant on how pagination is implemented)

I couldn't think how to do the check generically so left it out but may be worth revisiting. Possible a catch all on errors that says something like 'There was a problem transforming response data. Is the response paginated?' if it fails on an unpaginated Endpoint

File: django-testbed/frontend/src/models/User.js
Comment (davecoates) [reply django-testbed/frontend/src/models/User.js:35]: If not initially then long term I think yes

File: django-testbed/frontend/src/models/User.js
Comment (davecoates) [reply django-testbed/frontend/src/models/User.js:106]: There's 3 fields defined here to cover different use cases. `referredBy` is the simple use case:

```
referredBy: new IntegerField({
   asyncChoices: new AsyncChoices(asyncChoicesOptions),
 }),
```

in terms of reducing `asyncChoicesOptions` it's fairly generic so I imagine we'll have `ViewModelAsyncChoices` or something like that that does that (may be default for foreign key fields?)

File: js-packages/@prestojs/ui-antd/src/getWidgetForField.ts
Comment (davecoates) [reply js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:53]: no can do, prettier removes them

Comment (davecoates) [reply js-packages/@prestojs/ui-antd/src/getWidgetForField.ts:53]: https://github.com/prettier/prettier/issues/3805

File: js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx
Comment (davecoates) [reply js-packages/@prestojs/ui-antd/src/widgets/SelectAsyncChoiceWidget.tsx:18]: Reasonably specific to how antd select works. The usage of async choices is specific to ant but the AsyncChoices interface is generic (or at least that's the goal)

File: js-packages/@prestojs/util/src/Labeled.ts
Comment (davecoates) [reply js-packages/@prestojs/util/src/Labeled.ts]: I haven't seen `Rich` or `Plain`

The reason for the distinction is really for when you want to use string interpolation or other text transformation vs just render something

File: js-packages/@prestojs/util/src/identifiable.ts
Comment (davecoates) [reply js-packages/@prestojs/util/src/identifiable.ts]: I think it was really because `id` is the intuitive thing if implementing `Identifiable` it's just that it's not suitable for our use case (although would work most of the time). That's where I started from anyway but didn't want to force `ViewModel` to conform to this so added `_pk` as the support. 

I guess we could just change it to `_pk` it just sounds a bit more specific to a database implementation

(this isn't new behaviour in this PR BTW - I just refactored how the type was defined to make it easier to work with internally)

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:61]: oops I changed the name from `useListDeps` to `useListProps` but still use dependencies in the comment.

The props returned here is anything required by the eg. widget or hooks that use the async choices. So pagination is the common thing:

```ts
useListProps() {
  const paginator = usePaginator(endpoints.list);
  return { paginator };
}
```

It also can be used to transform the props passed in - so if you wanted to change how the keywords from a generic widget are transformed before they are passed through to the async function (eg. the endpoint) you can do this here without having to have a custom widget.

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:63]: See https://github.com/prestojs/prestojs/pull/67/files#diff-a1b1883790df4d0587b34a90efa4a851R120

You can return it from `listProps` otherwise it's passed straight through in `useAsyncChoices`

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:86]: Same kind of deal - most obvious usage I had for it was reading from cache always for existing values to avoid unnecessary fetches, eg https://github.com/prestojs/prestojs/pull/67/files#diff-013e735fdb71bd7fe985d673e4eb416cR62

File: js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:119]: No this is just for if a value is present but no label can be found

Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/fields/AsyncChoices.ts:119]: (widget handles it - there's a default on the widget or can be specified as a prop)

File: js-packages/@prestojs/util/src/Labeled.ts
Comment (davecoates) [reply js-packages/@prestojs/util/src/Labeled.ts:3]: Normally yes but currently it's failing to build

File: js-packages/@prestojs/util/src/Labeled.ts
Comment (levic) [reply js-packages/@prestojs/util/src/Labeled.ts]: Can we leave it up to the dev then whether it returns a string or a Node?

If they try to do string interpolation on a Node then they'll get an error, but they were the one who chose to return a Node so they can fix it.

Comment (davecoates) [reply js-packages/@prestojs/util/src/Labeled.ts]: You wouldn't get an error would just end up with a string like "Welcome [object Object]"

The intention was to provide different guarantees

Rich - suitable to be rendered to screen 
Plain - suitable to be used in an alert box, fill a text box, send via API call etc

You don't think there's advantage in separating them?

File: js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts
Comment (levic) [inline js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:686]: Is `_pk` still needed at all? Can we just replace it entirely with `_key`?

**Review by levic — COMMENTED** [review]
Looks good other than the question about _key / _pk: is _pk there just because of history or does it have a distinct meaning from _key? (esp does ViewModel really need both?)

File: js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts
Comment (davecoates) [reply js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:686]: Mainly because it was already there and used throughout + parallel to django's `record.pk`

i guess it would make sense to remove it and just use _key

File: js-packages/@prestojs/util/src/Labeled.ts
Comment (levic) [inline js-packages/@prestojs/util/src/Labeled.ts]: That should be `getTextLabel` for consistency with naming above

**Pull request merged**
